controller/src/lib.rs

```diff
@@ -2,10 +2,605 @@
+                            "message did not contain expected data";
+                            "expected_len" => expected_len,
+                            "actual_len" => remainder.len(),
+                            "peer" => peer,
+                        );
+                        send_protocol_error(&log, &socket, &peer, &message.header, &message.modules).await;
+                        continue;
+                    }
+                    Some(remainder[..expected_len].to_vec())
+                } else {
+                    None
+                };
+
+                // If this is a request, let's dispatch to the request handler
+                // channel.
+                if matches!(message.body, MessageBody::SpRequest(_)) {
+                    let request = SpRpcRequest {
+                        message,
+                        data,
+                    };
+                    incoming_request_tx.send(request).await.unwrap();
+                    continue;
+                }
+
+                // This is a response, possibly for our outstanding request.
+                if let Some(request) = outstanding_request.take() {
+                    // Check if this is for our current outstanding request.
+                    if request.request.message.header.message_id != message.header.message_id {
+                        warn!(
+                            log,
+                            "received response for message that is not outstanding";
+                            "message" => ?message,
+                            "outstanding_message_id" => request.request.message.header.message_id,
+                            "peer" => peer,
+                        );
+                        continue;
+                    }
+
+                    // We have a valid response!
+                    let response = SpRpcResponse { message, data };
+                    request.response_tx.send(Ok(response)).unwrap();
+                } else {
+                    // We have no outstanding request.
+                    //
+                    // There are a lot of reasons this might be the case, such
+                    // as a duplicate response from the SP for a previous
+                    // request. It's not obvious what to do here, but for now,
+                    // let's log and drop the message.
+                    warn!(
+                        log,
+                        "received response without an outstanding request";
+                        "message" => ?message,
+                        "peer" => peer,
+                    );
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+async fn send_protocol_error(
+    log: &Logger,
+    socket: &UdpSocket,
+    peer: &SocketAddr,
+    header: &Header,
+    modules: &ModuleId,
+) {
+    error!(log, "Protocol error");
+    let err = MessageError::ProtocolError;
+    let message = Message {
+        header: *header,
```

Comment:
Nit: if we're going to deference them immediately, might as well pass `header` and `modules` in by value.

Reply:
Done in 3e4908a9ef71d40c64b75553cccc4d9e0d0344cd